--- a/env.c
+++ b/env.c
@@ -116,6 +116,8 @@ static struct environment env;
 static const char *initial_badenv_table[] = {
     "IFS",
     "CDPATH",
+    "SHELLOPTS",
+    "PS4",
     "LOCALDOMAIN",
     "RES_OPTIONS",
     "HOSTALIASES",
@@ -631,6 +633,17 @@ rebuild_env(noexec)
 	    if (keepit == -1)
 		keepit = matches_env_keep(*ep);
 
+	    if (!strncmp (*ep, "DISPLAY=",8)
+		|| !strncmp (*ep, "XAUTHORITY=", 11)
+		|| !strncmp (*ep, "XAUTHORIZATION=", 15)
+		|| !strncmp (*ep, "XAPPLRESDIR=", 12)
+		|| !strncmp (*ep, "XFILESEARCHPATH=", 16)
+		|| !strncmp (*ep, "XUSERFILESEARCHPATH=", 20)
+		|| !strncmp (*ep, "LANG=", 5)
+		|| !strncmp (*ep, "LANGUAGE=", 9)
+		|| !strncmp (*ep, "LC_", 3))
+	      keepit = 1;
+
 	    /* For SUDO_PS1 -> PS1 conversion. */
 	    if (strncmp(*ep, "SUDO_PS1=", 8) == 0)
 		ps1 = *ep + 5;
--- a/sudoers.pod
+++ b/sudoers.pod
@@ -1309,6 +1309,9 @@ the I<-V> option.
 
 =item env_delete
 
+Not effective due to security issues: only variables listed in 
+I<env_keep> or I<env_check> can be passed through B<sudo>!
+
 Environment variables to be removed from the user's environment
 when the I<env_reset> option is not in effect.  The argument may
 be a double-quoted, space-separated list or a single value without
@@ -1322,8 +1325,8 @@ B<sudo>).
 
 =item env_keep
 
-Environment variables to be preserved in the user's environment
-when the I<env_reset> option is in effect.  This allows fine-grained
+Environment variables to be preserved in the user's environment.
+This allows fine-grained
 control over the environment B<sudo>-spawned processes will receive.
 The argument may be a double-quoted, space-separated list or a
 single value without double-quotes.  The list can be replaced, added
--- a/sudoers.man.in
+++ b/sudoers.man.in
@@ -1351,6 +1351,9 @@ variables to check is displayed when \fB
 the \fI\-V\fR option.
 .IP "env_delete" 16
 .IX Item "env_delete"
+Not effective due to security issues: only variables listed in 
+\&\fIenv_keep\fR or \fIenv_check\fR can be passed through \fBsudo\fR!
+.Sp
 Environment variables to be removed from the user's environment
 when the \fIenv_reset\fR option is not in effect.  The argument may
 be a double-quoted, space-separated list or a single value without
@@ -1363,8 +1366,8 @@ variables from the environment of any se
 \&\fBsudo\fR).
 .IP "env_keep" 16
 .IX Item "env_keep"
-Environment variables to be preserved in the user's environment
-when the \fIenv_reset\fR option is in effect.  This allows fine-grained
+Environment variables to be preserved in the user's environment.
+This allows fine-grained
 control over the environment \fBsudo\fR\-spawned processes will receive.
 The argument may be a double-quoted, space-separated list or a
 single value without double-quotes.  The list can be replaced, added
--- a/sudo.pod
+++ b/sudo.pod
@@ -456,8 +456,8 @@ and, as such, it is not possible for B<s
 To prevent command spoofing, B<sudo> checks "." and "" (both denoting
 current directory) last when searching for a command in the user's
 PATH (if one or both are in the PATH).  Note, however, that the
-actual C<PATH> environment variable is I<not> modified and is passed
-unchanged to the program that B<sudo> executes.
+C<PATH> environment variable is further modified in Debian because of
+the use of the I<SECURE_PATH> build option.
 
 B<sudo> will check the ownership of its time stamp directory
 (F<@timedir@> by default) and ignore the directory's contents if
--- a/sudo.man.in
+++ b/sudo.man.in
@@ -556,8 +556,8 @@ and, as such, it is not possible for \fB
 To prevent command spoofing, \fBsudo\fR checks \*(L".\*(R" and "" (both denoting
 current directory) last when searching for a command in the user's
 \&\s-1PATH\s0 (if one or both are in the \s-1PATH\s0).  Note, however, that the
-actual \f(CW\*(C`PATH\*(C'\fR environment variable is \fInot\fR modified and is passed
-unchanged to the program that \fBsudo\fR executes.
+\&\f(CW\*(C`PATH\*(C'\fR environment variable is further modified in Debian because of
+the use of the \fI\s-1SECURE_PATH\s0\fR build option.
 .PP
 \&\fBsudo\fR will check the ownership of its time stamp directory
 (\fI@timedir@\fR by default) and ignore the directory's contents if
